business_product_location/models/business_product_location.py

```diff
@@ -0,0 +1,160 @@
+
+class BusinessProductLine(models.Model):
+    _name = 'business.product.line'
+    _rec_name = "product_id"
+
+    @api.model
+    def _default_product_uom_id(self):
+        return self.env.ref('product.product_uom_unit')
+
+    product_id = fields.Many2one('product.product', string='Product',
+                                 required=True)
+    product_qty = fields.Float(
+        'Product Quantity', required=True,
+        digits_compute=dp.get_precision('Product Unit of Measure'))
+    product_uom_id = fields.Many2one('product.uom', 'Product Unit of Measure',
+                                     required=True,
+                                     default=_default_product_uom_id)
+    business_product_location_id = fields.Many2one(
+        'business.product.location', 'Parent business product location',
+        required=True)
+
+    @api.onchange('product_id')
+    def _onchange_product_id(self):
+        """ Change UoM if product_id changes
+        """
+        if self.product_id:
+            self.product_uom_id = self.product_id.uom_id
+
+    @api.onchange('product_uom_id')
+    def _onchange_product_uom_id(self):
+        """ Check the selected UoM with the product UoM
+        """
+        res = {}
+        if self.product_id and self.product_uom_id:
+            if self.product_id.uom_id.category_id.id != \
+                    self.product_uom_id.category_id.id:
+                res['warning'] = {
+                    'title': _('Warning'),
+                    'message': _('The Product Unit of Measure you chose '
+                                 'has a different category than in the '
+                                 'product form.')}
+                self.product_uom_id = self.product_id.uom_id
+
+
+class Product(models.Model):
+    _inherit = 'product.product'
+
+    business_usage_ids = fields.One2many('business.product.line', 'product_id',
+                                         'Business Usage')
+
+
+class StockLocation(models.Model):
+    _inherit = 'stock.location'
+
+    business_usage_id = fields.Many2one('business.product.location',
+                                        'Business Usage')
+
+
+class BusinessProductLocationReport(models.AbstractModel):
+    _name = 'report.business_product_location.report_bpl'
+
+    @api.multi
+    def render_html(self, data=None):
+        """ Get the list of product excepted in the location and the list of
+            product in the location with the quantity
+        """
+        docs = self.env['stock.location'].browse(self._ids)
+        product_obj = self.env['product.product']
+
+        less_product_by_loc = {}
+        more_product_by_loc = {}
+
+        for location in docs:
+            if location.business_usage_id:
+                location_name = '%s - %s' % (location.name,
+                                             location.business_usage_id.name)
+                more_product_by_loc[location_name] = {}
+                less_product_by_loc[location_name] = {}
```

Comment:
Using a string as dictionary key is most time a bad idea. Nothing makes you sure that you won't somehow have 2 identical tuples of (location.name, location.business_usage_id.name). Also, using a tuple of ids (location.id, location.business_usage_id.id) would be more efficient.